File: hash/max_good_subseqs_len/max_good_subseqs_len.py

def find_max_length(nums: list[int], k: int) -> int:  # LeetCode Q.3177.
    max_len = 1  # Base case.

    # ith idx stores best subseq len when using only nums[:i + 1].
    prev_diff_best, current_diff_best = [], []  # Only need latest 2 diffs.

    # At current diff, each num's best subseq len when served as subseq end.
    nums2lens: dict[int, int] = dict()
    for diff in range(k + 1):  # Differences go from 0 to k.
        for idx, num in enumerate(nums):
            if num in nums2lens.keys():  # Extend subseq w/o changing diff.
                nums2lens[num] += 1

            else:
                nums2lens.update({num: 1})  # Base case.

            if min(diff, idx) > 0:  # Can extend previous diff's earlier subseqs.
                if prev_diff_best[idx - 1] + 1 > nums2lens[num]:
                    nums2lens[num] = prev_diff_best[idx - 1] + 1

            if current_diff_best and nums2lens[num] <= current_diff_best[-1]:
                current_diff_best.append(current_diff_best[-1])

            else:
                current_diff_best.append(nums2lens[num])
                if nums2lens[num] > max_len:
                    max_len = nums2lens[num]

        prev_diff_best.clear()  # Reset for the next diff.
        prev_diff_best.extend(current_diff_best)
        current_diff_best.clear()
        nums2lens.clear()

    return max_len

File: hash/max_good_subseqs_len/max_good_subseqs_len.cpp
#include <vector>
#include <unordered_map>
using namespace std;

int maximumLength(vector<int> &nums, int k) // LeetCode Q.3177.
{
    int max_len = 1; // Base case.

    // ith idx stores best subseq len when using only nums[:i + 1].
    vector<int> prevDiffBest; // Only need latest 2 diffs.
    vector<int> currentDiffBest;

    // At current diff, each num's best subseq len when served as subseq end.
    unordered_map<int, int> nums2Lens;

    for (int diff = 0; diff < k + 1; diff++) // Differences go from 0 to k.
    {
        for (int idx = 0; idx < nums.size(); idx++)
        {
            if (nums2Lens.find(nums[idx]) != nums2Lens.end())
                nums2Lens[nums[idx]] += 1; // Extend subseq w/o changing diff.

            else
                nums2Lens[nums[idx]] = 1; // Base case.

            if (min(diff, idx) > 0) // Can extend previous diff's earlier subseqs.
            {
                if (1 + prevDiffBest[idx - 1] > nums2Lens[nums[idx]])
                    nums2Lens[nums[idx]] = 1 + prevDiffBest[idx - 1];
            }

            if (nums2Lens[nums[idx]] > max_len)
                max_len = nums2Lens[nums[idx]];

            if (idx >= 1 && currentDiffBest[idx - 1] >= nums2Lens[nums[idx]])
                currentDiffBest.push_back(currentDiffBest[idx - 1]);

            else
                currentDiffBest.push_back(nums2Lens[nums[idx]]);
        }

        prevDiffBest.clear(); // Reset for the next diff.

        move(
            currentDiffBest.begin(), currentDiffBest.end(),
            back_inserter(prevDiffBest));

        currentDiffBest.clear();
        nums2Lens.clear();
    }

    return max_len;
}